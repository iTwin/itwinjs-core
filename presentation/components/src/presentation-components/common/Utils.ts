/*---------------------------------------------------------------------------------------------
* Copyright (c) Bentley Systems, Incorporated. All rights reserved.
* See LICENSE.md in the project root for license terms and full copyright notice.
*--------------------------------------------------------------------------------------------*/
/** @packageDocumentation
 * @module Core
 */

import * as React from "react";
import { Descriptor, Field, LabelCompositeValue, LabelDefinition } from "@bentley/presentation-common";
import { Presentation } from "@bentley/presentation-frontend";
import { Primitives, PrimitiveValue, PropertyDescription, PropertyRecord, PropertyValueFormat } from "@bentley/ui-abstract";
import { IPropertyValueRenderer, PropertyValueRendererManager } from "@bentley/ui-components";
import { InstanceKeyValueRenderer } from "../properties/InstanceKeyValueRenderer";
import { FIELD_NAMES_SEPARATOR } from "./ContentBuilder";

/**
 * An interface of something that has a priority.
 * @internal
 */
export interface IPrioritized {
  priority: number;
}

/**
 * An interface of something that has a name.
 * @internal
 */
export interface INamed {
  name: string;
}

/**
 * A sorting algorithm for `Array.sort` that sorts items by priority and name.
 * Higher priority items appear first in the list. If priorities are equal, then
 * name property is used (in ascending order).
 *
 * @internal
 */
export const priorityAndNameSortFunction = (a: IPrioritized & INamed, b: IPrioritized & INamed): number => {
  if (a.priority > b.priority)
    return -1;
  if (a.priority < b.priority)
    return 1;
  return a.name.localeCompare(b.name);
};

const localizationNamespaceName = "PresentationComponents";

/**
 * Registers 'PresentationComponents' localization namespace and returns callback
 * to unregister it.
 * @internal
 */
export const initializeLocalization = async () => {
  await Presentation.i18n.registerNamespace(localizationNamespaceName).readFinished;
  return () => Presentation.i18n.unregisterNamespace(localizationNamespaceName);
};

/**
 * Registers custom property value renderers and returns cleanup callback that unregisters them.
 * @internal
 */
export const initializePropertyValueRenderers = async () => {
  const customRenderers: Array<{ name: string, renderer: IPropertyValueRenderer }> = [
    { name: "SelectableInstance", renderer: new InstanceKeyValueRenderer() },
  ];

  for (const { name, renderer } of customRenderers) {
    PropertyValueRendererManager.defaultManager.registerRenderer(name, renderer);
  }

  return () => {
    for (const { name } of customRenderers) {
      PropertyValueRendererManager.defaultManager.unregisterRenderer(name);
    }
  };
};

/**
 * Translate a string with the specified id from `PresentationComponents`
 * localization namespace. The `stringId` should not contain namespace - it's
 * prepended automatically.
 *
 * @internal
 */
export const translate = (stringId: string): string => {
  stringId = `${localizationNamespaceName}:${stringId}`;
  return Presentation.i18n.translate(stringId);
};

/**
 * Creates a display name for the supplied component
 * @internal
 */
export const getDisplayName = <P>(component: React.ComponentType<P>): string => {
  if (component.displayName)
    return component.displayName;
  if (component.name)
    return component.name;
  return "Component";
};

/**
 * Finds a field given the name of property record created from that field.
 * @internal
 */
export const findField = (descriptor: Descriptor, recordPropertyName: string): Field | undefined => {
  let fieldsSource: { getFieldByName: (name: string) => Field | undefined } | undefined = descriptor;
  const fieldNames = recordPropertyName.split(FIELD_NAMES_SEPARATOR);
  while (fieldsSource && fieldNames.length) {
    const field: Field | undefined = fieldsSource.getFieldByName(fieldNames.shift()!);
    fieldsSource = (field && field.isNestedContentField()) ? field : undefined;
    if (!fieldNames.length)
      return field;
  }
  return undefined;
};

/**
 * Creates property record for label using label definition.
 * @internal
 */
export const createLabelRecord = (label: LabelDefinition, name: string): PropertyRecord => {
  const value: PrimitiveValue = {
    displayValue: label.displayValue,
    value: createPrimitiveLabelValue(label),
    valueFormat: PropertyValueFormat.Primitive,
  };
  const property: PropertyDescription = {
    displayLabel: "Label",
    typename: label.typeName,
    name,
  };

  return new PropertyRecord(value, property);
};

const createPrimitiveLabelValue = (label: LabelDefinition) => {
  return LabelDefinition.isCompositeDefinition(label) ? createPrimitiveCompositeValue(label.rawValue) : label.rawValue;
};

const createPrimitiveCompositeValue = (compositeValue: LabelCompositeValue): Primitives.Composite => {
  return {
    separator: compositeValue.separator,
    parts: compositeValue.values.map((part) => ({
      displayValue: part.displayValue,
      typeName: part.typeName,
      rawValue: createPrimitiveLabelValue(part),
    })),
  };
};
