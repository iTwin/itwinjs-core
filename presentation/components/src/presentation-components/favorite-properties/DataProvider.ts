/*---------------------------------------------------------------------------------------------
* Copyright (c) Bentley Systems, Incorporated. All rights reserved.
* See LICENSE.md in the project root for license terms and full copyright notice.
*--------------------------------------------------------------------------------------------*/
/** @packageDocumentation
 * @module FavoriteProperties
 */

import { PropertyData } from "@itwin/components-react";
import { Id64Arg, using } from "@itwin/core-bentley";
import { IModelConnection } from "@itwin/core-frontend";
import { CategoryDescription, KeySet, Ruleset } from "@itwin/presentation-common";
import { createSelectionScopeProps, Presentation } from "@itwin/presentation-frontend";
import { translate } from "../common/Utils";
import { PresentationPropertyDataProvider } from "../propertygrid/DataProvider";

/** @internal */
export const FAVORITES_CATEGORY_NAME = "Favorite";

/** @internal */
export const getFavoritesCategory = (): CategoryDescription => {
  return {
    name: FAVORITES_CATEGORY_NAME,
    label: translate("categories.favorite.label"),
    description: translate("categories.favorite.description"),
    priority: Number.MAX_VALUE,
    expand: true,
  };
};

/**
 * An data provider interface for returning favorite properties for the given elements
 * @public
 */
export interface IFavoritePropertiesDataProvider {
  /** Returns property data for an element. */
  getData: (imodel: IModelConnection, elementIds: Id64Arg | KeySet) => Promise<PropertyData>;
}

/**
 * Props for [[FavoritePropertiesDataProvider]]
 * @public
 */
export interface FavoritePropertiesDataProviderProps {
  /**
   * Id of the ruleset to use when requesting properties or a ruleset itself. If not
   * set, default presentation rules are used which return content for the selected elements.
   */
  ruleset?: Ruleset | string;

  /** @internal */
  propertyDataProviderFactory?: (imodel: IModelConnection, ruleset?: Ruleset | string) => PresentationPropertyDataProvider;
}

/**
 * Presentation Rules-driven element favorite properties data provider implementation.
 * @public
 */
export class FavoritePropertiesDataProvider implements IFavoritePropertiesDataProvider {

  private _customRuleset?: Ruleset | string;
  private _propertyDataProviderFactory: (imodel: IModelConnection, ruleset?: Ruleset | string) => PresentationPropertyDataProvider;

  /**
   * Should fields with no values be included in the property list. No value means:
   * - For *primitive* fields: null, undefined, "" (empty string)
   * - For *array* fields: [] (empty array)
   * - For *struct* fields: {} (object with no members)
   */
  public includeFieldsWithNoValues: boolean;

  /**
   * Should fields with composite values be included in the property list.
   * Fields with composite values:
   * - *array* fields.
   * - *struct* fields.
   */
  public includeFieldsWithCompositeValues: boolean;

  /** Constructor. */
  constructor(props?: FavoritePropertiesDataProviderProps) {
    this.includeFieldsWithNoValues = true;
    this.includeFieldsWithCompositeValues = true;
    this._customRuleset = /* istanbul ignore next */ props?.ruleset;
    this._propertyDataProviderFactory = props && props.propertyDataProviderFactory
      ? props.propertyDataProviderFactory
      : /* istanbul ignore next */ (imodel: IModelConnection, ruleset?: Ruleset | string) => {
        const provider = new PresentationPropertyDataProvider({ imodel, ruleset });
        provider.isNestedPropertyCategoryGroupingEnabled = false;
        return provider;
      };
  }

  /**
   * Returns PropertyData for the specified elements.
   * PropertyData only contains a single category for favorite properties (if there are any).
   */
  public async getData(imodel: IModelConnection, elementIds: Id64Arg | KeySet): Promise<PropertyData> {
    if (elementIds instanceof KeySet) {
      return using(this._propertyDataProviderFactory(imodel, this._customRuleset), async (propertyDataProvider) => {
        propertyDataProvider.keys = elementIds;
        propertyDataProvider.includeFieldsWithNoValues = this.includeFieldsWithNoValues;
        propertyDataProvider.includeFieldsWithCompositeValues = this.includeFieldsWithCompositeValues;
        const propertyData = await propertyDataProvider.getData();

        // leave only favorite properties
        const favoritesCategory = getFavoritesCategory();
        propertyData.categories = propertyData.categories.filter((c) => c.name === favoritesCategory.name);
        propertyData.records = propertyData.records.hasOwnProperty(favoritesCategory.name) ?
          { [favoritesCategory.name]: propertyData.records[favoritesCategory.name] } : {};
        return propertyData;
      });
    }

    const keys = await Presentation.selection.scopes.computeSelection(imodel, elementIds, createSelectionScopeProps(Presentation.selection.scopes.activeScope));
    return this.getData(imodel, keys);
  }
}
