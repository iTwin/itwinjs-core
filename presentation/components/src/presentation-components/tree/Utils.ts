/*---------------------------------------------------------------------------------------------
* Copyright (c) Bentley Systems, Incorporated. All rights reserved.
* See LICENSE.md in the project root for license terms and full copyright notice.
*--------------------------------------------------------------------------------------------*/
/** @packageDocumentation
 * @module Tree
 */

import { LabelDefinition, Node, NodeKey, PartialNode, PageOptions as PresentationPageOptions } from "@bentley/presentation-common";
import { PropertyRecord } from "@bentley/ui-abstract";
import { DelayLoadedTreeNodeItem, ItemColorOverrides, ItemStyle, PageOptions as UiPageOptions } from "@bentley/ui-components";
import { CheckBoxState } from "@bentley/ui-core";
import { StyleHelper } from "../common/StyleHelper";
import { createLabelRecord } from "../common/Utils";

/** @internal */
export const PRESENTATION_TREE_NODE_KEY = "__presentation-components/key";

/** @internal */
export interface CreateTreeNodeItemProps {
  appendChildrenCountForGroupingNodes?: boolean;
}

/** @internal */
export function createTreeNodeItems(
  nodes: ReadonlyArray<Readonly<Node>>,
  parentId?: string,
  props?: CreateTreeNodeItemProps,
): DelayLoadedTreeNodeItem[] {
  const list = new Array<DelayLoadedTreeNodeItem>();
  for (const node of nodes)
    list.push(createTreeNodeItem(node, parentId, props));
  return list;
}

/** @internal */
export function createTreeNodeItem(
  node: Readonly<Node>,
  parentId?: string,
  props?: CreateTreeNodeItemProps,
): DelayLoadedTreeNodeItem {
  const item: DelayLoadedTreeNodeItem = {
    id: createTreeNodeId(node.key),
    label: createNodeLabelRecord(node, !!props?.appendChildrenCountForGroupingNodes),
  };
  assignOptionalTreeNodeItemFields(item, node, parentId);
  return item;
}

/** @internal */
export function createPartialTreeNodeItem(
  node: PartialNode,
  parentId: string | undefined,
  props: CreateTreeNodeItemProps,
): Partial<DelayLoadedTreeNodeItem> {
  const item: Partial<DelayLoadedTreeNodeItem> = {};
  if (node.key !== undefined) {
    item.id = createTreeNodeId(node.key);
    item.label = createNodeLabelRecord(node, !!props.appendChildrenCountForGroupingNodes);
  }

  assignOptionalTreeNodeItemFields(item, node, parentId);
  return item;
}

/** @internal */
export function createTreeNodeId(key: NodeKey): string {
  return [...key.pathFromRoot].reverse().join("/");
}

function assignOptionalTreeNodeItemFields(
  item: Partial<DelayLoadedTreeNodeItem>,
  node: Partial<Node>,
  parentId?: string,
): void {
  (item as any)[PRESENTATION_TREE_NODE_KEY] = node.key;

  if (parentId) {
    item.parentId = parentId;
  }

  if (node.description) {
    item.description = node.description;
  }

  if (node.hasChildren) {
    item.hasChildren = true;
  }

  if (node.isExpanded) {
    item.autoExpand = true;
  }

  if (node.imageId) {
    item.icon = node.imageId;
  }

  if (node.isCheckboxVisible) {
    item.isCheckboxVisible = true;
    if (node.isChecked) {
      item.checkBoxState = CheckBoxState.On;
    }

    if (!node.isCheckboxEnabled) {
      item.isCheckboxDisabled = true;
    }
  }

  const style = createTreeNodeItemStyle(node);
  if (Object.keys(style).length > 0) {
    item.style = style;
  }

  if (node.extendedData) {
    item.extendedData = node.extendedData;
  }
}

function createTreeNodeItemStyle(node: Partial<Node>): ItemStyle {
  const style: ItemStyle = {};
  if (StyleHelper.isBold(node)) {
    style.isBold = true;
  }

  if (StyleHelper.isItalic(node)) {
    style.isItalic = true;
  }

  const colorOverrides: ItemColorOverrides = {};
  const foreColor = StyleHelper.getForeColor(node);
  if (foreColor) {
    colorOverrides.color = foreColor;
  }

  const backColor = StyleHelper.getBackColor(node);
  if (backColor) {
    colorOverrides.backgroundColor = backColor;
  }

  if (Object.keys(colorOverrides).length > 0) {
    style.colorOverrides = colorOverrides;
  }

  return style;
}

/** @internal */
export function pageOptionsUiToPresentation(pageOptions?: UiPageOptions): PresentationPageOptions | undefined {
  if (pageOptions)
    return { ...pageOptions };
  return undefined;
}

function createNodeLabelRecord(node: Node, appendChildrenCountForGroupingNodes: boolean): PropertyRecord {
  let labelDefinition = node.label;
  if (appendChildrenCountForGroupingNodes && NodeKey.isGroupingNodeKey(node.key)) {
    const countDefinition: LabelDefinition = {
      displayValue: `(${node.key.groupedInstancesCount})`,
      rawValue: `(${node.key.groupedInstancesCount})`,
      typeName: "string",
    };
    labelDefinition = {
      displayValue: `${labelDefinition.displayValue} ${countDefinition.displayValue}`,
      rawValue: {
        separator: " ",
        values: [
          labelDefinition,
          countDefinition,
        ],
      },
      typeName: LabelDefinition.COMPOSITE_DEFINITION_TYPENAME,
    };
  }
  return createLabelRecord(labelDefinition, "node_label");
}
