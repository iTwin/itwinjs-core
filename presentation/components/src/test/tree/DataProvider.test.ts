/*---------------------------------------------------------------------------------------------
* Copyright (c) Bentley Systems, Incorporated. All rights reserved.
* See LICENSE.md in the project root for license terms and full copyright notice.
*--------------------------------------------------------------------------------------------*/

import "@bentley/presentation-frontend/lib/test/_helpers/MockFrontendEnvironment";
import { expect } from "chai";
import * as faker from "faker";
import * as sinon from "sinon";
import { BeEvent, Logger } from "@bentley/bentleyjs-core";
import { IModelConnection } from "@bentley/imodeljs-frontend";
import { Node, RegisteredRuleset } from "@bentley/presentation-common";
import * as moq from "@bentley/presentation-common/lib/test/_helpers/Mocks";
import { PromiseContainer, ResolvablePromise } from "@bentley/presentation-common/lib/test/_helpers/Promises";
import {
  createRandomECInstancesNode, createRandomECInstancesNodeKey, createRandomNodePathElement, createRandomRuleset,
} from "@bentley/presentation-common/lib/test/_helpers/random";
import { Presentation, PresentationManager, RulesetManager, RulesetVariablesManager } from "@bentley/presentation-frontend";
import { PageOptions } from "@bentley/ui-components";
import { PresentationTreeDataProvider } from "../../presentation-components/tree/DataProvider";
import { pageOptionsUiToPresentation } from "../../presentation-components/tree/Utils";
import { createRandomTreeNodeItem } from "../_helpers/UiComponents";

describe("TreeDataProvider", () => {

  let rulesetId: string;
  let provider: PresentationTreeDataProvider;
  let onVariableChanged: BeEvent<(variableId: string) => void>;
  const presentationManagerMock = moq.Mock.ofType<PresentationManager>();
  const rulesetVariablesManagerMock = moq.Mock.ofType<RulesetVariablesManager>();
  const imodelMock = moq.Mock.ofType<IModelConnection>();

  before(() => {
    rulesetId = faker.random.word();
    Presentation.setPresentationManager(presentationManagerMock.object);
  });

  after(() => {
    Presentation.terminate();
  });

  beforeEach(() => {
    onVariableChanged = new BeEvent();
    presentationManagerMock.reset();
    rulesetVariablesManagerMock.reset();
    presentationManagerMock.setup((x) => x.vars(moq.It.isAny())).returns(() => rulesetVariablesManagerMock.object);
    rulesetVariablesManagerMock.setup((x) => x.onVariableChanged).returns(() => onVariableChanged);
    provider = new PresentationTreeDataProvider({ imodel: imodelMock.object, ruleset: rulesetId });
  });

  describe("dispose", () => {

    it("disposes registered ruleset", async () => {
      const registerPromise = new ResolvablePromise<RegisteredRuleset>();
      const rulesetsManagerMock = moq.Mock.ofType<RulesetManager>();
      rulesetsManagerMock.setup(async (x) => x.add(moq.It.isAny())).returns(async () => registerPromise);
      presentationManagerMock.setup((x) => x.rulesets()).returns(() => rulesetsManagerMock.object);

      const ruleset = await createRandomRuleset();
      const p = new PresentationTreeDataProvider({ imodel: imodelMock.object, ruleset });
      const rulesetDisposeSpy = sinon.spy();
      await registerPromise.resolve(new RegisteredRuleset(ruleset, "test", rulesetDisposeSpy));

      expect(rulesetDisposeSpy).to.not.be.called;
      p.dispose();
      expect(rulesetDisposeSpy).to.be.calledOnce;
    });

  });

  describe("rulesetId", () => {

    it("returns rulesetId provider is initialized with", () => {
      expect(provider.rulesetId).to.eq(rulesetId);
    });

  });

  describe("imodel", () => {

    it("returns imodel provider is initialized with", () => {
      expect(provider.imodel).to.eq(imodelMock.object);
    });

  });

  describe("getNodesCount", () => {

    it("returns presentation manager result through `getNodesCount` when page options not set", async () => {
      const parentKey = createRandomECInstancesNodeKey();
      const parentNode = createRandomTreeNodeItem(parentKey);
      presentationManagerMock
        .setup(async (x) => x.getNodesCount({ imodel: imodelMock.object, rulesetOrId: rulesetId, parentKey }))
        .returns(async () => 999)
        .verifiable();
      const actualResult = await provider.getNodesCount(parentNode);
      expect(actualResult).to.eq(999);
      presentationManagerMock.verifyAll();
    });

    it("returns presentation manager result through `getNodesAndCount` when page options are set", async () => {
      const resultNodes = [createRandomECInstancesNode(), createRandomECInstancesNode()];
      const parentKey = createRandomECInstancesNodeKey();
      const parentNode = createRandomTreeNodeItem(parentKey);
      presentationManagerMock
        .setup(async (x) => x.getNodesAndCount({ imodel: imodelMock.object, rulesetOrId: rulesetId, paging: { start: 0, size: 123 }, parentKey }))
        .returns(async () => ({ nodes: resultNodes, count: resultNodes.length }))
        .verifiable();
      provider.pagingSize = 123;
      const actualResult = await provider.getNodesCount(parentNode);
      expect(actualResult).to.eq(resultNodes.length);
      presentationManagerMock.verifyAll();
    });

    it("memoizes result", async () => {
      const parentKeys = [createRandomECInstancesNodeKey(), createRandomECInstancesNodeKey()];
      const parentNodes = parentKeys.map((key) => createRandomTreeNodeItem(key));
      const resultContainers = [new PromiseContainer<{ nodes: Node[], count: number }>(), new PromiseContainer<{ nodes: Node[], count: number }>()];

      presentationManagerMock
        .setup(async (x) => x.getNodesAndCount({ imodel: imodelMock.object, rulesetOrId: rulesetId, paging: { start: 0, size: 10 }, parentKey: parentKeys[0] }))
        .returns(async () => resultContainers[0].promise)
        .verifiable(moq.Times.once());
      presentationManagerMock
        .setup(async (x) => x.getNodesAndCount({ imodel: imodelMock.object, rulesetOrId: rulesetId, paging: { start: 0, size: 10 }, parentKey: parentKeys[1] }))
        .returns(async () => resultContainers[1].promise)
        .verifiable(moq.Times.once());

      provider.pagingSize = 10;
      const promises = [
        provider.getNodesCount(parentNodes[0]),
        provider.getNodesCount(parentNodes[0]),
        provider.getNodesCount(parentNodes[1]),
      ];
      resultContainers.forEach((c, index) => c.resolve({ nodes: [createRandomECInstancesNode(), createRandomECInstancesNode()], count: index }));
      const results = await Promise.all(promises);
      expect(results[0]).to.eq(results[1]).to.eq(0);
      expect(results[2]).to.eq(1);

      presentationManagerMock.verifyAll();
    });

    it("clears memoized result when ruleset variables changes", async () => {
      const parentKey = createRandomECInstancesNodeKey();
      const parentNode = createRandomTreeNodeItem(parentKey);
      presentationManagerMock
        .setup(async (x) => x.getNodesAndCount({ imodel: imodelMock.object, rulesetOrId: rulesetId, paging: { start: 0, size: 10 }, parentKey }))
        .returns(async () => ({ nodes: [createRandomECInstancesNode(), createRandomECInstancesNode()], count: 2 }))
        .verifiable(moq.Times.exactly(2));

      provider.pagingSize = 10;
      await provider.getNodesCount(parentNode);
      onVariableChanged.raiseEvent("testVar");
      await provider.getNodesCount(parentNode);

      presentationManagerMock.verifyAll();
    });

    it("uses `getNodesCount` data source override if supplied", async () => {
      const override = sinon.mock().resolves(0);
      provider = new PresentationTreeDataProvider({ imodel: imodelMock.object, ruleset: rulesetId, dataSourceOverrides: { getNodesCount: override } });
      await provider.getNodesCount(undefined);
      // eslint-disable-next-line deprecation/deprecation
      presentationManagerMock.verify(async (x) => x.getNodesCount(moq.It.isAny(), moq.It.isAny()), moq.Times.never());
      expect(override).to.be.calledOnce;
    });

  });

  describe("getNodes", () => {

    it("returns presentation manager result", async () => {
      const parentKey = createRandomECInstancesNodeKey();
      const parentNode = createRandomTreeNodeItem(parentKey);
      const pageOptions: PageOptions = { start: 0, size: faker.random.number() };
      presentationManagerMock
        .setup(async (x) => x.getNodesAndCount({ imodel: imodelMock.object, rulesetOrId: rulesetId, paging: pageOptionsUiToPresentation(pageOptions), parentKey }))
        .returns(async () => ({ nodes: [createRandomECInstancesNode(), createRandomECInstancesNode()], count: 2 }))
        .verifiable();
      const actualResult = await provider.getNodes(parentNode, pageOptions);
      expect(actualResult).to.matchSnapshot();
      presentationManagerMock.verifyAll();
    });

    it("memoizes result", async () => {
      const parentKeys = [createRandomECInstancesNodeKey(), createRandomECInstancesNodeKey()];
      const parentNodes = parentKeys.map((key) => createRandomTreeNodeItem(key));
      const resultNodesFirstPageContainer0 = new PromiseContainer<{ nodes: Node[], count: number }>();
      const resultNodesFirstPageContainer1 = new PromiseContainer<{ nodes: Node[], count: number }>();
      const resultNodesNonFirstPageContainer = new PromiseContainer<{ nodes: Node[], count: number }>();

      presentationManagerMock
        .setup(async (x) => x.getNodesAndCount({ imodel: imodelMock.object, rulesetOrId: rulesetId, paging: undefined, parentKey: parentKeys[0] }))
        .returns(async () => resultNodesFirstPageContainer0.promise)
        .verifiable(moq.Times.once());
      presentationManagerMock
        .setup(async (x) => x.getNodesAndCount({ imodel: imodelMock.object, rulesetOrId: rulesetId, paging: { start: 0, size: 0 }, parentKey: parentKeys[0] }))
        .verifiable(moq.Times.never());
      presentationManagerMock
        .setup(async (x) => x.getNodesAndCount({ imodel: imodelMock.object, rulesetOrId: rulesetId, paging: { start: 1, size: 0 }, parentKey: parentKeys[0] }))
        .returns(async () => resultNodesNonFirstPageContainer.promise)
        .verifiable(moq.Times.once());
      presentationManagerMock
        .setup(async (x) => x.getNodesAndCount({ imodel: imodelMock.object, rulesetOrId: rulesetId, paging: { start: 0, size: 1 }, parentKey: parentKeys[1] }))
        .returns(async () => resultNodesFirstPageContainer1.promise)
        .verifiable(moq.Times.once());

      const promises = [
        provider.getNodes(parentNodes[0], undefined), provider.getNodes(parentNodes[0], undefined),
        provider.getNodes(parentNodes[0], { start: 0, size: 0 }), provider.getNodes(parentNodes[0], { start: 0, size: 0 }),
        provider.getNodes(parentNodes[0], { start: 1, size: 0 }), provider.getNodes(parentNodes[0], { start: 1, size: 0 }),
        provider.getNodes(parentNodes[1], { start: 0, size: 1 }), provider.getNodes(parentNodes[1], { start: 0, size: 1 }),
      ];
      resultNodesFirstPageContainer0.resolve({ nodes: [createRandomECInstancesNode()], count: 1 });
      resultNodesFirstPageContainer1.resolve({ nodes: [createRandomECInstancesNode()], count: 1 });
      resultNodesNonFirstPageContainer.resolve({ nodes: [createRandomECInstancesNode()], count: 1 });
      const results = await Promise.all(promises);

      expect(results[0]).to.eq(results[1], "results[0] should eq results[1]");
      expect(results[2])
        .to.eq(results[3], "results[2] should eq results[3]")
        .to.eq(results[0], "both results[2] and results[3] should eq results[0]");
      expect(results[4]).to.eq(results[5], "results[4] should eq results[5]");
      expect(results[6]).to.eq(results[7], "results[6] should eq results[7]");

      presentationManagerMock.verifyAll();
    });

    it("uses `getNodesAndCount` data source override if supplied", async () => {
      const override = sinon.mock().resolves({ count: 0, nodes: [] });
      provider = new PresentationTreeDataProvider({ imodel: imodelMock.object, ruleset: rulesetId, dataSourceOverrides: { getNodesAndCount: override } });
      await provider.getNodes();
      presentationManagerMock.verify(async (x) => x.getNodesAndCount(moq.It.isAny()), moq.Times.never());
      expect(override).to.be.calledOnce;
    });

    it("logs a warning when requesting nodes and pagingSize is not the same as passed pageOptions", async () => {
      const pageOptions: PageOptions = { start: 0, size: 10 };
      const loggerSpy = sinon.spy(Logger, "logWarning");
      const result = { nodes: [createRandomECInstancesNode(), createRandomECInstancesNode()], count: 2 };
      presentationManagerMock.setup(async (x) => x.getNodesAndCount(moq.It.isAny())).returns(async () => result);

      // Paging size is not set and pageOptions are passed
      await provider.getNodes(undefined, pageOptions);
      expect(loggerSpy.calledOnce).to.be.true;
      loggerSpy.resetHistory();

      // Paging size is set and no pageOptions are passed
      provider.pagingSize = 10;
      await provider.getNodes();
      expect(loggerSpy.notCalled).to.be.true;
      loggerSpy.resetHistory();

      // Paging size is set and pageOptions are passed but not equal to paging size
      provider.pagingSize = 20;
      await provider.getNodes(undefined, pageOptions);
      expect(loggerSpy.calledOnce).to.be.true;
      loggerSpy.resetHistory();

      // Paging size is set and pageOptions are passed and equal to paging size
      provider.pagingSize = 10;
      await provider.getNodes(undefined, pageOptions);
      expect(loggerSpy.notCalled).to.be.true;
    });

  });

  describe("getFilteredNodes", () => {

    it("returns presentation manager result", async () => {
      const filter = faker.random.word();
      presentationManagerMock
        .setup(async (x) => x.getFilteredNodePaths({ imodel: imodelMock.object, rulesetOrId: rulesetId }, filter))
        .returns(async () => [createRandomNodePathElement(), createRandomNodePathElement()])
        .verifiable();
      const actualResult = await provider.getFilteredNodePaths(filter);
      expect(actualResult).to.matchSnapshot();
      presentationManagerMock.verifyAll();
    });

    it("uses `getFilteredNodePaths` data source override if supplied", async () => {
      const override = sinon.mock().resolves([]);
      provider = new PresentationTreeDataProvider({ imodel: imodelMock.object, ruleset: rulesetId, dataSourceOverrides: { getFilteredNodePaths: override } });
      await provider.getFilteredNodePaths("test");
      presentationManagerMock.verify(async (x) => x.getFilteredNodePaths(moq.It.isAny(), moq.It.isAny()), moq.Times.never());
      expect(override).to.be.calledOnce;
    });

  });

});
