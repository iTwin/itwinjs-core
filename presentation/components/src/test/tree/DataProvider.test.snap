// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TreeDataProvider getFilteredNodes returns presentation manager result 1`] = `
Array [
  Object {
    "children": Array [],
    "index": 644,
    "isMarked": true,
    "node": Object {
      "backColor": "rgb(
    215,
    125,
    116
  )",
      "description": "Rerum laborum ipsa.",
      "foreColor": "#6EE32D",
      "hasChildren": true,
      "imageId": "Soft",
      "isCheckboxEnabled": false,
      "isCheckboxVisible": true,
      "isChecked": true,
      "isEditable": true,
      "isExpanded": false,
      "isSelectionDisabled": false,
      "key": Object {
        "instanceKeys": Array [
          Object {
            "className": "SQL",
            "id": "0x42e20000015b9b",
          },
          Object {
            "className": "viral",
            "id": "0x1b78000000b915",
          },
        ],
        "pathFromRoot": Array [
          "05b9034c-bb1b-47e5-a34b-ee893267e90a",
          "8528b854-ec02-4dd8-bfe4-a098a1d96013",
        ],
        "type": "ECInstancesNode",
        "version": 2,
      },
      "label": Object {
        "displayValue": "Auto Loan Account",
        "rawValue": "Awesome",
        "typeName": "string",
      },
    },
  },
  Object {
    "children": Array [],
    "index": 668,
    "node": Object {
      "backColor": undefined,
      "description": "Autem ipsa vel voluptatem quia recusandae.",
      "foreColor": "#2F04EF",
      "hasChildren": false,
      "imageId": "Chips",
      "isCheckboxEnabled": true,
      "isCheckboxVisible": false,
      "isChecked": false,
      "isEditable": false,
      "isExpanded": false,
      "isSelectionDisabled": false,
      "key": Object {
        "instanceKeys": Array [
          Object {
            "className": "Applications",
            "id": "0x6b0b000000b7ad",
          },
          Object {
            "className": "EXE",
            "id": "0xc6b4000001712e",
          },
        ],
        "pathFromRoot": Array [
          "a3944426-64d1-4202-bc22-44b3f7075dc2",
          "845797cd-fdf3-4462-ba83-b9a1ecd911f2",
        ],
        "type": "ECInstancesNode",
        "version": 2,
      },
      "label": Object {
        "displayValue": "Cordoba Oro",
        "rawValue": "purple",
        "typeName": "string",
      },
    },
  },
]
`;

exports[`TreeDataProvider getNodes returns presentation manager result 1`] = `
Array [
  Object {
    "description": "Aut et tempore voluptatum rerum unde autem nulla.",
    "hasChildren": true,
    "icon": "Dobra",
    "id": "d1ea9e80-e8b1-4710-a1b1-7fa1c654a1f8/4a55713b-f10a-423a-a6e0-c54b1b677e78",
    "key": Object {
      "instanceKeys": Array [
        Object {
          "className": "payment",
          "id": "0x16be80000004405",
        },
        Object {
          "className": "Sleek Granite Soap",
          "id": "0x91bb0000011198",
        },
      ],
      "pathFromRoot": Array [
        "4a55713b-f10a-423a-a6e0-c54b1b677e78",
        "d1ea9e80-e8b1-4710-a1b1-7fa1c654a1f8",
      ],
      "type": "ECInstancesNode",
      "version": 2,
    },
    "label": PropertyRecord {
      "property": Object {
        "displayLabel": "Label",
        "name": "node_label",
        "typename": "string",
      },
      "value": Object {
        "displayValue": "Turkish Lira",
        "value": "Handcrafted Granite Chicken",
        "valueFormat": 0,
      },
    },
    "parentId": "62a2f68d-cb37-4531-995a-ac9b559e9ef8",
  },
  Object {
    "autoExpand": true,
    "description": "Quia nihil molestias delectus necessitatibus reiciendis consequatur.",
    "hasChildren": true,
    "id": "b8a8c076-222c-4b3c-9c60-4ee0adecf300/ea29ec34-9504-448a-93b5-6ec63a5c5450",
    "isCheckboxDisabled": true,
    "isCheckboxVisible": true,
    "key": Object {
      "instanceKeys": Array [
        Object {
          "className": "optical",
          "id": "0x5a610000016c10",
        },
        Object {
          "className": "Senegal",
          "id": "0xb5ae0000000db4",
        },
      ],
      "pathFromRoot": Array [
        "ea29ec34-9504-448a-93b5-6ec63a5c5450",
        "b8a8c076-222c-4b3c-9c60-4ee0adecf300",
      ],
      "type": "ECInstancesNode",
      "version": 2,
    },
    "label": PropertyRecord {
      "property": Object {
        "displayLabel": "Label",
        "name": "node_label",
        "typename": "string",
      },
      "value": Object {
        "displayValue": "heuristic",
        "value": "Avon",
        "valueFormat": 0,
      },
    },
    "parentId": "62a2f68d-cb37-4531-995a-ac9b559e9ef8",
    "style": Object {
      "colorOverrides": Object {
        "color": 2186652,
      },
    },
  },
]
`;
